### PR TITLE
Error when assertion function calls aren't CFA'd

### LINKED ISSUE #33580: CFA Assertions Not Active on Implicitly-Typed Function Calls
**TypeScript Version:**  3.7.0-insiders.20190923 (this [tgz](https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/45201/artifacts?artifactName=tgz&fileId=FCC96CD1470D020E21159D18DAC814941830D31C28D43629A094AF099B11388B02&fileName=/typescript-3.7.0-insiders.20190923.tgz) from [this comment](https://github.com/microsoft/TypeScript/pull/32695#issuecomment-534269544))

**Search Terms:** `asserts` `cfa` `control flow analysis`

**Code**

```ts
class Foo {
}

interface BarChild {
    parent: Bar;
}

class Bar {
    static staticAdd(f: Foo): asserts f is Foo & BarChild {}
    add(f: Foo): asserts f is Foo & BarChild {}
    innerTest() {
        const innerFoo = new Foo();
        this.add(innerFoo);
        innerFoo.parent;  // this compiles successfully
    }
}

function main() {
    const foo1 = new Foo();
    const bar1 = new Bar();
    bar1.add(foo1);
    foo1.parent;  // this is an ERROR - Property 'parent' does not exist on type 'Foo'. ts(2339)

    const foo2 = new Foo();
    Bar.staticAdd(foo2);
    foo2.parent;  // this compiles successfully
}
```

The `Bar.add` assertion works when called inside another method of `Bar` but fails when called externally from `main`. I added a static function assertion to show that assertion still works otherwise in `main`.

**Expected behavior:** In `main` after `bar1.add(foo1)`, `foo1` should be of type `Foo & BarChild`.

**Actual behavior:** In `main` after `bar1.add(foo1)`, `foo1` is still type `Foo`.

**Playground Link:** I've duplicated the sample code above in [this](https://www.typescriptlang.org/play/?ts=Nightly#code/MYGwhgzhAEBiD29oG8BQBfVqCWA7ALgKYBOAZmMIdAEJjEDCAFtiACYqrRfQAOdhBAFw06AbgxZQkGLWIdu0CPjD5swRctXAAgq1YAKUsITwAlMOkl8MUtGwwT0AGQiGzNikwKwew8cTm0JbE1tC29nCIzq5MLOzIXtx4uCQAKoRK+qbyCtzA8LhKdrgpxI4AvNApAO6R8Fniudz4zBAAdD4GySQmpo1NxaUmbXzEAviiXAD0U9AtEfkAtjwsGYoArsCUUKTrICAAnpzcmJiou7jAqgXQi2B4WTlc+YX4YYgAjNCVNXUNx88CkUAEZ0L4-Qi1WT-BSg4gfDq+UifPoA97wBGjcaTaAzOatOwwMC4aAAUQASuSAPLk6AAWmgAAViPAeFYDtAAORYgic6CseBrXDwN6EAAe9jeN3wBzZXJMnLacwg+gATABmdUATlMWAULyKyPgqu+VUhf1RClkbSUKjUugMRtVlu4TpG-AIOLx8xgSxWIDWEE22wgu32R3QQA) playground link, but at the time of posting the nightly version does not have these changes.

**Related Issues:** https://github.com/microsoft/TypeScript/pull/32695 This feature was merged to master yesterday, so instead of commenting on a closed PR I opened a new issue.

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we issue errors when calls to assertion functions are written in a manner that isn't recognized by control flow analysis (see #32695 for the specific requirements). For example:

```ts
class Test {
    assert(value: unknown): asserts value {}
}

function f20(x: unknown) {
    const assert1 = (value: unknown): asserts value => {}
    assert1(typeof x === "string");  // Error
    const assert2: (value: unknown) => asserts value = () => {};
    assert2(typeof x === "string");  // Ok
    const t1 = new Test();
    t1.assert(typeof x === "string");  // Error
    const t2: Test = new Test();
    t2.assert(typeof x === "string");  // Ok
    const getAssert = () => assert1;
    getAssert()(typeof x === "string");  // Error
}
```

The reported errors are:

t.ts(7,5): error TS2775: Assertions require every name in the call target to be declared with an explicit type annotation.

t.ts(11,5): error TS2775: Assertions require every name in the call target to be declared with an explicit type annotation.

t.ts(15,5): error TS2776: Assertions require the call target to be an identifier or qualified name.

Fixes #33580.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot run dt
@typescript-bot test this

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2726,6 +2726,14 @@
         "category": "Error",
         "code": 2774
     },
+    "Control flow effects of calls to assertion and never-returning functions are reflected only when every variable or property referenced in the function expression is declared with an explicit type annotation.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: I don't know what this is telling a user. Is it something like the following?

```
This assertion call cannot be properly analyzed because not every variable and property used in the current function has an explicit type annotation.

This 'never'-returning function call cannot be properly analyzed because not every variable and property used in the current function has an explicit type annotation.
```

Along with

```
This assertion call cannot be properly analyzed because the function being called is not a qualified name.

This 'never'-returning function call cannot be properly analyzed because the function being called is not a qualified name.
```

File: tests/baselines/reference/assertionTypePredicates1.errors.txt [hunk 2]
```diff
@@ -147,4 +150,21 @@ tests/cases/conformance/controlFlow/assertionTypePredicates1.ts(124,15): error T
                   ~~~~~~~~~~~~~~~~~~~~~~
 !!! error TS1228: A type predicate is only allowed in return type position for functions and methods.
     }
+    
+    function f20(x: unknown) {
+        const assert = (value: unknown): asserts value => {}
+        assert(typeof x === "string");  // Error
+        ~~~~~~
+!!! error TS2775: Control flow effects of calls to assertion and never-returning functions are reflected only when every variable or property referenced in the function expression is declared with an explicit type annotation.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/assertionTypePredicates1.errors.txt]: *Which* variable doesn't have an annotation? Can you give a related span?

**Comment by DanielRosenwasser** [comment]
@typescript-bot user test this
@typescript-bot pack this

**Comment by jack-williams** [comment]
A couple of candidate messages.

```ts
assert1(typeof x === "string");  // Error
// Assertion calls must be explicitly typed.
// Identifier 'assert1' lacks an explicit type annotation and is not subject to control flow analysis.
```

```ts
getAssert()(typeof x === "string");  // Error
// Assertion calls must be explicitly resolved.
// Function expression is not an identifier or qualified-name and is not subject to control flow analysis.
```

File: src/compiler/diagnosticMessages.json (hunk 1 above)
Comment (sandersn) [reply src/compiler/diagnosticMessages.json]: 1. I think that "explicitly never-returning function" is really just another kind of assertion. We could introduce the concept of *guard* = "explicitly never-returning function" if it's too ambiguous. 
2. Mentioning control flow isn't necessary for fixing the error -- and presumably most people understand the concept of assertion better than control flow anyway.
3. As for @DanielRosenwasser's wording: "This assertion call" is implied by the error span. The fact that it's an assertion is implied by the error span+error wording. So neither of those is necessary. Wording this categorically also allows us to flip the message back to positive.

```
Assertions must be an identifier or qualified name.
```


```
Assertions must reference a declaration with an explicit type annotation.
  Related: This is the declaration that should be annotated.
```

4. We know that the *expression* has type annotations because `hasTypePredicateOrNeverReturnType(signature)` is true.
5. If we want to handle the case where there is more than one declaration, the message can be

```
Assertions must reference declaration(s) with an explicit type annotation.
  Related: This is one declaration that must be annotated.
  Related: This is one declaration that must be annotated.
  :  
  : 
```

**Comment by sandersn** [comment]
Oh, I didn't read @jack-williams comments until after commenting. Basically, what he said. =)

**Comment by DanielRosenwasser** [comment]
> Mentioning control flow isn't necessary for fixing the error

Agreed!

> "This assertion call" is implied by the error span.

Yes and no - I find it easier to just have the error message tell me what's wrong rather than reconstructing the intent from the span as well.

> Wording this categorically also allows us to flip the message back to positive.

The suggestions both of you have are much more digestible, and made it clear to me exactly what the error check is. I only have nits at this point. :)

What do you think of:

> Assertion functions can only be called when they were declared with an explicit type annotation.
> Related: This is the declaration that must be annotated.

and

> Assertion functions can only be called when referenced by an identifier or qualified name.

?

**Comment by sandersn** [comment]
No, I like the shorter wording better. The longer message only improves the precision of the wording, but won't help people (especially naive people) track down the problem.

1. "A declaration with an explicit type annotation" is shorter and therefore clearer than "When they [functions] were declared with an explicit type annotation", and much more flexible because it's a noun phrase. Just use it as the object of the sentence!
2. I guess that you ended up with "Assertion functions" because of "can only be called", which you got because of "when they were declared ...". Whether that reconstruction is right or not, it makes it sound like "an assertion function" could be different from "an assertion", but we don't actually have those two concepts. So we might as well call the one concept we do have "an assertion".

**Comment by DanielRosenwasser** [comment]
Any way to get a related span in? Still not clear what declaration needs to be annotated.

File: tests/baselines/reference/assertionTypePredicates1.errors.txt [hunk 3]
```diff
@@ -147,4 +150,23 @@ tests/cases/conformance/controlFlow/assertionTypePredicates1.ts(124,15): error T
                   ~~~~~~~~~~~~~~~~~~~~~~
 !!! error TS1228: A type predicate is only allowed in return type position for functions and methods.
     }
+    
+    function f20(x: unknown) {
+        const assert = (value: unknown): asserts value => {}
+        assert(typeof x === "string");  // Error
+        ~~~~~~
+!!! error TS2775: Assertions require every name in the call target to be declared with an explicit type annotation.
+!!! related TS2728 tests/cases/conformance/controlFlow/assertionTypePredicates1.ts:128:11: 'assert' is declared here.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/assertionTypePredicates1.errors.txt]: `'{0}' must be annotated` would be a lot nicer 🙂

**Comment by ahejlsberg** [comment]
Latest commits exclude never-returning functions from check that they match CFA requirements and improve error messages, per discussion in today's design meeting.

File: tests/baselines/reference/assertionTypePredicates1.errors.txt (hunk 3 above)
Comment (ahejlsberg) [reply tests/baselines/reference/assertionTypePredicates1.errors.txt]: It would, but do we really want yet another diagnostic?

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -4062,6 +4062,12 @@ namespace ts {
         return node.kind === SyntaxKind.Identifier || isPropertyAccessEntityNameExpression(node);
     }
 
+    export function isDottedName(node: Expression): boolean {
```
Comment (weswigham) [inline src/compiler/utilities.ts]: BTW, will this need to be updated for optional chains? Should a `t1?.assert(b)` be a thing?

File: tests/baselines/reference/assertionTypePredicates1.errors.txt (hunk 3 above)
Comment (DanielRosenwasser) [reply tests/baselines/reference/assertionTypePredicates1.errors.txt]: I know you already merged this, but for the future, diagnostics are cheap! They bear very little overhead for the <20 of us on the core team compared to at least hundreds of thousands of TypeScript users.

Comment (ahejlsberg) [reply tests/baselines/reference/assertionTypePredicates1.errors.txt]: We discussed in the team room and agreed that the existing message is fine, so I didn't change it.

**Comment by mgcrea** [comment]
Strangely, with 3.8-beta (or 3.7-stable), this will trigger the `Assertions require every name in the call target to be declared with an explicit type annotation. ts(2775)` error:

```ts
export const assert = (value: unknown, message?: string): asserts value => {
  if (!value) {
    throw new AssertionError({message});
  }
};
```

While this does work properly:

```ts
export const assert: (value: unknown, message?: string) => asserts value = (value: unknown, message?: string) => {
  if (!value) {
    throw new AssertionError({message});
  }
};
```

Is this working as intended?
```
